### PR TITLE
kill global g_graphicsContext, g_textureManager, and g_LargeTextureManager

### PR DESCRIPTION
**Pull request by FernetMenta**
- move GraphicsContext from guilib to windowing
- make GraphicsContext a member of CWinSystemBase
- kill g_textureManager
- kill g_LargeTextureManager

There is more work to be done in this area. All remaining dependencies from windowing to guilib have to be cut.

### CODE REVIEW

**Comment by FernetMenta** [comment]
due to segfault on exit and failing test suit I had to kill 2 more globals

File: xbmc/guilib/GUIComponent.cpp [hunk 1]
```diff
@@ -20,13 +20,17 @@
...
 #include "dialogs/GUIDialogYesNo.h"
+#include "GUILargeTextureManager.h"
 #include "ServiceBroker.h"
 #include "URL.h"
 
 CGUIComponent::CGUIComponent()
 {
   m_pWindowManager.reset(new CGUIWindowManager());
+  m_pTextureManager.reset(new CGUITextureManager());
+  m_pLargeTextureManager.reset(new CGUILargeTextureManager());
```
Comment (garbear) [inline xbmc/guilib/GUIComponent.cpp]: if `m_pWindowManager->Initialize()` doesn't depend on these then they should happen after m_pWindowManager is initialized

File: xbmc/guilib/GUIComponent.cpp [hunk 2]
```diff
@@ -35,13 +39,27 @@ CGUIComponent::~CGUIComponent()
 {
   CServiceBroker::UnregisterGUI();
   m_pWindowManager->DeInitialize();
+  
+  m_pWindowManager.reset();
+  m_pTextureManager.reset();
+  m_pLargeTextureManager.reset();
```
Comment (garbear) [inline xbmc/guilib/GUIComponent.cpp]: Can you deinitialize these in reverse order?

Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: if this matters, something is really wrong and we should fix the root cause.

Comment (garbear) [reply xbmc/guilib/GUIComponent.cpp]: If we reverse them, it will show us if anything is wrong so we can fix it.

Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: deinit is reverse order is happy path. doing it differently challenges the system more.

Comment (garbear) [reply xbmc/guilib/GUIComponent.cpp]: huh? isn't fixing the system worth the effort?

Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: Yes ofc, this is why changing the order here should not matter. Keeping the order in a sequence that is worse, will bring issues to surface faster

Comment (garbear) [reply xbmc/guilib/GUIComponent.cpp]: I don't understand, isn't RAII-inspired order, where resources are unacquired in the opposite order, better than what is here now?

Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: Right, uninit in reverse order is better in general. But the order should not matter here. If it matters I want to know and fix the actual issue.

Comment (garbear) [reply xbmc/guilib/GUIComponent.cpp]: It matters from a maintainability standpoint. It's easier for a programmer to verify the order is reversed properly than try to reason about the intention of breaking convention.

Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: Ok, then I will remove those lines completely. The only reason they are pointers is to avoid includes in header file

Comment (garbear) [reply xbmc/guilib/GUIComponent.cpp]: Nonexistent code is easiest to maintain :)

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/guilib/GUIComponent.cpp (hunk 1 above)
Comment (FernetMenta) [reply xbmc/guilib/GUIComponent.cpp]: I dropped m_pWindowManager->Initialize() from ctor. Putting behaviour in ctor is not a good style anyway.

File: xbmc/windowing/GraphicContext.cpp [hunk 3]
```diff
@@ -1089,7 +1089,7 @@ float CGraphicContext::GetFPS() const
 {
   if (m_Resolution != RES_INVALID)
   {
-    RESOLUTION_INFO info = g_graphicsContext.GetResInfo();
+    RESOLUTION_INFO info = CServiceBroker::GetWinSystem().GetGfxContext().GetResInfo();
```
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/windowing/GraphicContext.cpp]: right, thanks.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@DaveTBlake why don't you try your testbuild proposal on this one.

**Comment by DaveTBlake** [comment]
Sure @FernetMenta ping me once it is on the mirrors and I will happily test it on Windows

**Comment by DaveTBlake** [comment]
@FernetMenta I have tried the gfxContext build that is on the mirrors http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20180403-ef9fae4c-gfxcontext-x64.exe.  It does not work,  crashing before showing the splash screen.

**Comment by FernetMenta** [comment]
This branch did not include afedchin's fix. I rebased. Please try again.

**Comment by DaveTBlake** [comment]
I think jenkins failed to build the wrong version. I have queued a win64 test build for it

**Comment by DaveTBlake** [comment]
@FernetMenta happy to confirm that win64 build works in simple testing - no crashes on entry or exit.

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
Since this seems to cure the crash on exit that master  currently has in Win64 it would be nice to merge this @FernetMenta

**Comment by DaveTBlake** [comment]
@FernetMenta could you rebase this please

**Comment by afedchin** [comment]
@DaveTBlake woulkd you like to merge this? it's possible without rebase.

**Comment by DaveTBlake** [comment]
Merge without rebase is a new concept for me @afedchin , what happens to the conflicts?

**Comment by a1rwulf** [comment]
@DaveTBlake If you "git merge" you'll get a conflict like you do when "git rebase". I've never tried to use this github web resolver and I'm not sure we want it, although the conflicts seem to be just in the header section.
The conflict resolution afaik is then part of the merge commit.

**Comment by Rechi** [comment]
Don't resolve the conflicts in the merge commit, just wait till this PR is rebased.

**Comment by FernetMenta** [comment]
rebased

**Comment by DaveTBlake** [comment]
Retested on win64, seems fine, no crashing on start or exit. 
Will let jenkins complete building (again?) before merge

**Comment by DaveTBlake** [comment]
Odd, the win64 build failed yet was fine when  jenkins created the test build. One of those things?

**Pull request merged**